File: modules/agent/src/models/Lead.py
from sqlalchemy import Column, Text, DateTime, BigInteger, ForeignKey, func
from sqlalchemy.orm import relationship
from models import Base

"""Lead model (base table for Joined Table Inheritance)."""




class Lead(Base):
    """Lead SQLAlchemy model (base table for Joined Table Inheritance)."""

    __tablename__ = "Lead"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    tenant_id = Column(BigInteger, ForeignKey("Tenant.id", ondelete="CASCADE"), nullable=True)
    account_id = Column(BigInteger, ForeignKey("Account.id", ondelete="SET NULL"), nullable=True)
    deal_id = Column(BigInteger, ForeignKey("Deal.id", ondelete="CASCADE"), nullable=False)
    type = Column(Text, nullable=False)  # Discriminator: 'Job', 'Opportunity', 'Partnership', etc.
    title = Column(Text, nullable=False)
    description = Column(Text, nullable=True)
    source = Column(Text, nullable=True)  # inbound/referral/event/campaign/agent/manual/etc.
    current_status_id = Column(BigInteger, ForeignKey("Status.id", ondelete="SET NULL"), nullable=True)
    owner_individual_id = Column(BigInteger, ForeignKey("Individual.id", ondelete="SET NULL"), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    tenant = relationship("Tenant", back_populates="leads")
    account = relationship("Account", back_populates="leads")
    deal = relationship("Deal", back_populates="leads")
    current_status = relationship("Status", back_populates="leads", foreign_keys=[current_status_id])
    projects = relationship("Project", secondary="LeadProject", back_populates="leads")

    # Joined table inheritance relationships
    job = relationship("Job", back_populates="lead", uselist=False)
    opportunity = relationship("Opportunity", back_populates="lead", uselist=False)
    partnership = relationship("Partnership", back_populates="lead", uselist=False)

File: modules/agent-go/internal/models/Lead.go
package models

import "time"

type Lead struct {
	ID                int64     `gorm:"primaryKey;autoIncrement" json:"id"`
	TenantID          *int64    `gorm:"index" json:"tenant_id"`
	AccountID         *int64    `gorm:"index" json:"account_id"`
	DealID            int64     `gorm:"not null" json:"deal_id"`
	Type              string    `gorm:"not null" json:"type"` // Job, Opportunity, Partnership
	Description       *string   `json:"description"`
	Source            *string   `json:"source"` // inbound, referral, event, campaign, agent, manual
	CurrentStatusID   *int64    `json:"current_status_id"`
	OwnerIndividualID *int64    `json:"owner_individual_id"`
	CreatedAt         time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt         time.Time `gorm:"autoUpdateTime" json:"updated_at"`
	CreatedBy         int64     `gorm:"not null" json:"created_by"`
	UpdatedBy         int64     `gorm:"not null" json:"updated_by"`

	// Relations
	CurrentStatus *Status `gorm:"foreignKey:CurrentStatusID" json:"current_status,omitempty"`
	Account       *Account    `gorm:"foreignKey:AccountID" json:"account,omitempty"`
	Projects      []Project   `gorm:"many2many:Lead_Project;joinForeignKey:lead_id;joinReferences:project_id" json:"projects,omitempty"`
}

func (Lead) TableName() string {
	return "Lead"
}
